### PR TITLE
[core] Prefer breaking lines at zero width space (ZWSP)

### PR DESCRIPTION
**Pull request by alexshalamov**
Port of https://github.com/mapbox/mapbox-gl-js/pull/8255

### CODE REVIEW

**Comment by tmpsantos** [comment]
@alexshalamov can we bump gl-js to get the render tests?

**Comment by alexshalamov** [comment]
@tmpsantos I can bump in another PR. mapbox/mapbox-gl-js#8255 does not add new render tests, just unit tests.

File: src/mbgl/text/shaping.cpp [hunk 1]
```diff
@@ -241,7 +247,9 @@ std::set<std::size_t> determineLineBreaks(const TaggedString& logicalInput,
     
     std::list<PotentialBreak> potentialBreaks;
     float currentX = 0;
-    
+    // Find first occurance of zero width space (ZWSP) character.
+    const bool hasServerSuggestedBreaks = logicalInput.rawText().find_first_of(u'\u200b') !=  std::string::npos;
```
Comment (pozdnyakov) [inline src/mbgl/text/shaping.cpp]: `const auto ZWSP = u'\u200b';` ?

File: src/mbgl/text/shaping.cpp [hunk 2]
```diff
@@ -256,11 +264,14 @@ std::set<std::size_t> determineLineBreaks(const TaggedString& logicalInput,
...
-        if ((i < logicalInput.length() - 1) &&
-            (util::i18n::allowsWordBreaking(codePoint) || util::i18n::allowsIdeographicBreaking(codePoint))) {
-            potentialBreaks.push_back(evaluateBreak(i+1, currentX, targetWidth, potentialBreaks,
-                                                    calculatePenalty(codePoint, logicalInput.getCharCodeAt(i+1)),
-                                                    false));
+        if (i < logicalInput.length() - 1) {
+            const bool allowsIdeographicBreak = util::i18n::allowsIdeographicBreaking(codePoint);
+            if (allowsIdeographicBreak || util::i18n::allowsWordBreaking(codePoint)) {
+                const bool penalizableIdeographicBreak = allowsIdeographicBreak && hasServerSuggestedBreaks;
+                potentialBreaks.push_back(evaluateBreak(i+1, currentX, targetWidth, potentialBreaks,
```
Comment (pozdnyakov) [inline src/mbgl/text/shaping.cpp]: nit: `i + 1`

Comment (pozdnyakov) [reply src/mbgl/text/shaping.cpp]: `size_t nextIndex = i + 1;`

[the remark this answers is not in the document]
Comment (alexshalamov) [reply src/mbgl/text/shaping.cpp]: done

**Pull request merged**